virtool/uploads/api.py

```diff
@@ -53,59 +58,136 @@ async def naive_writer(req, file_id):
-            "$set": {
+            document.update({
                 "size": size,
-                "ready": True
+                "name_on_disk": f"{upload_id}-{name}",
+                "uploaded_at": uploaded_at
+            })
+
+            logger.debug(f"Upload succeeded: {upload_id}")
+
+            headers = {
+                "Location": f"/api/uploads/{upload_id}"
             }
-        })
 
-        logger.debug(f"Upload succeeded: {file_id}")
+            return json_response(document, status=201, headers=headers)
+        except asyncio.CancelledError:
+            logger.debug(f"Upload aborted: {upload_id}")
+
+            await session.delete(result)
+            await session.commit()
+
+            return aiohttp.web.Response(status=499)
+
+
+@routes.get("/api/uploads")
+async def find(req):
+    db = req.app["postgres"]
+    document = list()
+    filters = list()
+    user = req.query.get("user")
+
+    if user:
+        filters.append(Upload.user == user)
+
+    try:
+        filters.append(Upload.type == req.query["type"])
+    except KeyError:
+        pass
+
+    async with AsyncSession(db) as session:
+        if filters:
+            result = await session.execute(select(Upload).filter(*filters))
+        else:
+            result = await session.execute(select(Upload))
+
+        for upload in result.scalars().all():
+            document.append(upload.to_dict())
+
+    resp = json_response(document)
+
+    return resp
+
+
+@routes.get("/api/uploads/{id}")
+async def get(req):
+    db = req.app["postgres"]
+    upload_id = int(req.match_info["id"])
+
+    async with AsyncSession(db) as session:
+        result = (await session.execute(select(Upload).filter_by(id=upload_id))).scalar()
+
+        if not result:
+            return not_found("Upload record not found")
+
+    # check if the file has been removed as a result of a `DELETE` request
+    if result.removed:
+        return not_found("Uploaded file has already been removed")
+
+    upload_path = Path(req.app["settings"]["data_path"]) / "files" / result.name_on_disk
+
+    # check if the file has been manually removed by the user
+    if not upload_path.exists():
+        return not_found("Uploaded file not found at expected location")
+
+    return aiohttp.web.FileResponse(upload_path)
+
+
+@routes.delete("/api/uploads/{id}", permission="remove_file")
+async def delete(req):
+    db = req.app["postgres"]
+    upload_id = int(req.match_info["id"])
+
+    async with AsyncSession(db) as session:
+        result = (await session.execute(select(Upload).where(Upload.id == upload_id))).scalar()
+
+        if not result:
+            return not_found("Upload record not found")
+
+        if result.removed:
+            return bad_request("Uploaded file has already been removed")
 
-        headers = {
-            "Location": f"/api/files/{file_id}"
-        }
+        try:
+            os.remove(Path(req.app["settings"]["data_path"]) / "files" / result.name_on_disk)
```

Comment:
Need to use executor for this because it will block the server significantly if the filesystem is slow.

```py
await app["run_in_thread"](
  os.remove,
  path
)
```